ctapipe/reco/shower_reconstructor.py

```diff
@@ -0,0 +1,165 @@
+import astropy.units as u
+import numpy as np
+from astropy.coordinates import AltAz, SkyCoord
+from iminuit import Minuit
+from scipy.stats import poisson
+
+from ctapipe.containers import ReconstructedGeometryContainer
+from ctapipe.coordinates import CameraFrame, TelescopeFrame
+from ctapipe.core.traits import Unicode
+from ctapipe.image import GaussianShowermodel, ShowermodelPredictor
+from ctapipe.reco import Reconstructor
+
+
+class ShowerReconstructor(Reconstructor):
+    geometry_seed = Unicode(default_value="HillasReconstructor").tag(config=True)
+
+    def __init__(self, subarray, **kwargs):
+        super().__init__(subarray=subarray, **kwargs)
+
+    def __call__(self, event):
+        if self.geometry_seed not in event.dl2.stereo.geometry:
+            raise ValueError()
+        # fit shower parameters using likelihood
+        self.event = event  # ugly, only for _likelihood
+
+        # get all telescope properties that we need for the ShowermodelPredictor
+        tel_positions = {}
+        tel_solid_angles = {}
+        tel_mirror_area = {}
+        for tel_id in event.dl1.tel.keys():
+            tel_positions[tel_id] = self.subarray.positions[tel_id]
+            geometry = self.subarray.tel[tel_id].camera.geometry
+            tel_solid_angles[tel_id] = geometry.transform_to(TelescopeFrame()).pix_area
+            tel_mirror_area[tel_id] = self.subarray.tel[tel_id].optics.mirror_area
+
+        self.tel_positions = tel_positions
+        self.tel_solid_angles = tel_solid_angles
+        self.tel_mirror_area = tel_mirror_area
+
+        shower_parameters, errors = self._fit(event)
+        # store results in event structure
+        event.dl2.stereo.geometry[
+            self.__class__.__name__
+        ] = ReconstructedGeometryContainer(
+            alt=shower_parameters["altitude"],
+            az=shower_parameters["azimuth"],
+            alt_uncert=errors["altitude"],
+            az_uncert=errors["azimuth"],
+            core_x=shower_parameters["x"],
+            core_y=shower_parameters["y"],
+            core_tilted_x=None,
+            core_tilted_y=None,
+            h_max=shower_parameters["first_interaction"]
+            - shower_parameters["width"]
+            / 2
+            * np.cos(90 - shower_parameters["altitude"]),
+            average_intensity=shower_parameters["total_photons"],
+            telescopes=list(event.dl1.tel.keys()),
+            is_valid=True,
+            prefix=self.__class__.__name__,
+        )
+
+    def _fit(self, event):
+        # just call iminiut with likelihood and some seeds
+        # may also need limits
+        seeds = self._seeds(event)
+        minimizer = Minuit(self._likelihood, *seeds)
+        minimizer.limits = [
```

Comment:
Hard coded limits. Also, much more readable would be assigning like this:

```
minimizer.limits["x"] = (...,.... )
```

Reply:
Yes I know. This is still work in progress and not working correctly

Reply:
Might want to take a look at ImPACTReconstructor, as it also does likelihood minimization. Perhaps we can extract a common way to do it to avoid repetition.

Reply:
This should be pretty similar now. Generate images -> calculate likelihood -> fit using minuit with seeds from previous reconstructor